### PR TITLE
feat(firewall): payload sizing, byte budgets, tiktoken counter, honest summaries

### LINKED ISSUE #174: Surface truncation explicitly in summaries and fix boolean handling in `summarize()`
## Summary

Make the summarizer's data loss visible and correct: report how many facts/rows were
omitted when `max_facts`/row caps truncate output, and stop treating `bool` values
as numeric when computing column statistics.

## Why this matters

An LLM reading a summary Frame has no way to know it is seeing 20 facts out of
12,000 — silent truncation invites confidently wrong conclusions, which is an
agent-safety issue as much as a UX one. Separately, because `bool` is a subclass of
`int` in Python, boolean columns are averaged as numbers (e.g., "mean of `is_active`
= 0.7"), producing nonsense statistics in table summaries.

## Current evidence

- `firewall/summarize.py:11-26`: `summarize(data, *, max_facts: int = 20)` returns "(≤ max_facts)" facts with no omission marker; strings and reprs are truncated to 200 chars silently.
- `firewall/summarize.py:56`: numeric-column detection uses `isinstance(r.get(k), (int, float))` — `True`/`False` pass this check.
- `models.py` `Frame` has no field conveying totals vs shown counts (handles carry totals, but summaries do not reference them).

## External context

Not required for this issue.

## Proposed implementation

1. Append a final fact (and/or structured warning per ISSUE 25) when truncation
   occurs: "Showing 20 of N facts; full data available via handle <id>."
2. In numeric detection, exclude bools: `isinstance(v, (int, float)) and not
   isinstance(v, bool)`; summarize boolean columns as true/false counts instead.
3. Mark string truncation with an ellipsis indicator and a code.
4. Keep output deterministic (AGENTS.md requirement).

## AI-agent execution notes

- Inspect first: `firewall/summarize.py` (all `_summarize_*` helpers), `firewall/transform.py` (how facts reach Frames), `tests/test_firewall.py`.
- Determinism: counts and ordering must be stable.
- Edge cases: exactly `max_facts` items (no false truncation marker); mixed bool/int columns; empty datasets.
- Existing fact strings are asserted in tests — adjust tests deliberately, not by weakening.

## Acceptance criteria

- Truncated summaries state shown-vs-total counts; untruncated summaries carry no marker.
- Boolean columns are never averaged; they summarize as counts.
- Determinism preserved (same input → same facts).

## Test plan

Unit tests for truncation marker boundaries and bool columns in the summarize tests.
Run `make ci`.

## Documentation plan

Update `docs/context_firewall.md` summarization section; CHANGELOG `Fixed`/`Added`.

## Migration and compatibility notes

Summary text changes (new trailing fact); consumers parsing fact strings (not
contract) may need updates. Not expected to require migration.

## Risks and tradeoffs

Slightly longer summaries (one fact). Changing fact text can churn downstream
snapshot tests — call it out in the changelog.

## Suggested labels

product, reliability, good-first-ai-issue

### LINKED ISSUE #207: Avoid serializing entire payloads to measure size in the firewall hot path
## Summary

Replace the `len(json.dumps(data, default=str))` size estimation in
`firewall/transform.py` with an allocation-free size estimator (or a cached/
single-pass measurement), so every invocation does not pay a full JSON serialization
of the raw result purely to estimate its size.

## Why this matters

The firewall runs on *every* invoke. For large tool results (the exact case budgets
exist for — multi-MB query results), `json.dumps` doubles peak memory and adds
CPU proportional to payload size, before any truncation happens. The measurement
exists only to drive budget decisions, so an estimator with bounded error (or an
early-exit "bigger than budget" check) achieves the same decision at a fraction of
the cost.

## Current evidence

- `firewall/transform.py:125`: `raw_size = len(json.dumps(data, default=str))` — full serialization per transform.
- `firewall/budgets.py` consumes the size for mode-downgrade decisions; exact byte counts are not semantically required, thresholds are.
- `firewall/token_counting.py` defines the pluggable counting seam this estimator could share.

## External context

Not required for this issue.

## Proposed implementation

1. Implement a recursive size estimator that walks the structure summing
   approximate contributions (string lengths, numeric constants, container
   overhead) with an early exit once the relevant threshold is exceeded —
   thresholds are what budgets actually compare against.
2. Keep exact behavior at the decision level: golden tests asserting identical
   mode-downgrade decisions across a corpus of payloads (small, boundary, huge).
3. Measure before/after on a multi-MB synthetic payload; record numbers in the PR.
4. Consider sharing the walk with redaction (single traversal) as a follow-up only —
   do not couple the two in this change.

## AI-agent execution notes

- Inspect first: `firewall/transform.py` (all uses of `raw_size`), `firewall/budgets.py` (thresholds), `firewall/token_counting.py`, `tests/test_firewall.py`.
- Boundary risk: payloads near a threshold may flip decisions if the estimator diverges from `json.dumps` semantics — calibrate the estimator or keep exact counting under a size floor.
- Determinism required.
- Edge cases: non-JSON-serializable objects currently handled by `default=str`; datetimes; deeply nested structures (respect recursion limits).

## Acceptance criteria

- No full `json.dumps` of raw data solely for size measurement on the transform path.
- Mode-downgrade decisions are unchanged across the golden corpus, including boundary cases.
- Demonstrated CPU/memory improvement on a large payload (numbers in PR).

## Test plan

Golden decision tests; a perf check script; existing firewall tests. Run `make ci`.

## Documentation plan

Note estimation semantics in `docs/context_firewall.md` budgets section; CHANGELOG
`Changed` (performance).

## Migration and compatibility notes

Behavior-preserving at the decision level; not expected to require migration.

## Risks and tradeoffs

Estimator drift vs serialization is the main risk — bounded by golden tests and an
exactness floor. Slight code complexity increase in a security-adjacent module;
keep the estimator pure and well-tested.

## Suggested labels

performance, architecture, testing

### LINKED ISSUE #211: Add byte-size-aware budgeting and eviction to HandleStore
## Summary

Extend `HandleStore` beyond its entry-count cap (`max_entries=10_000`) with an
optional total-bytes budget and per-entry size cap, evicting oldest-first, so
10,000 multi-megabyte raw datasets cannot exhaust process memory.

## Why this matters

The store holds *raw, pre-firewall* datasets (by design — expansion needs them).
Entry count is a poor proxy for memory: one deployment's 10k entries are kilobytes,
another's are gigabytes. Since handle data is the most sensitive data in the
process (unredacted), bounding its residency is also a security-hygiene win:
less raw data held, less raw data exposable by any future bug.

## Current evidence

- `handles.py:17-32`: documented eviction design caps entries (`max_entries=10_000`, `_EVICT_INTERVAL=128`) but tracks no sizes.
- `kernel/_invoke.py` stores `raw_result.data` wholesale per invocation.
- No configuration exists for per-entry or aggregate byte limits.

## External context

Size-aware LRU/FIFO caches are the standard pattern for payload stores.

## Proposed implementation

1. Estimate entry size at `store()` time — reuse ISSUE 37's estimator (sequence
   after it, or implement the estimator here and share).
2. Add `max_total_bytes: int | None` and `max_entry_bytes: int | None`; on
   overflow evict oldest until within budget; oversized single entries are stored
   truncated-with-flag or rejected with a stable warning (decide; document).
3. Track and expose current usage (pairs with ISSUE 28 counters).
4. Keep `None` defaults so current behavior is unchanged until configured.

## AI-agent execution notes

- Inspect first: `handles.py` (store/evict paths, `_EVICT_INTERVAL`), `kernel/_invoke.py` (store call), `tests/test_handles.py`.
- Determinism: eviction order deterministic (insertion order).
- Edge cases: evicting a handle that an agent still references (expansion must fail with the existing expired/missing-handle error, not a crash); size accounting on eviction; zero/negative budgets rejected with typed errors.
- Don't change TTL semantics.

## Acceptance criteria

- With a byte budget configured, total estimated residency never exceeds it (tested with synthetic large entries).
- Expanding an evicted handle fails with the existing typed not-found/expired error.
- Default (None) behavior is unchanged.

## Test plan

Unit tests for budget enforcement, eviction order, oversized-entry policy, and
post-eviction expansion errors. Run `make ci`.

## Documentation plan

`docs/architecture.md` handle section + sizing guidance; CHANGELOG `Added`.

## Migration and compatibility notes

Opt-in; not expected to require migration.

## Risks and tradeoffs

Earlier eviction means more "handle expired" experiences for agents — surface
clear errors and document tuning. Size estimation cost per store is the price of
boundedness (cheap with ISSUE 37's estimator).

## Suggested labels

performance, reliability, security

### LINKED ISSUE #218: Implement the reserved `tiktoken` extra for accurate token counting
## Summary

Ship the `tiktoken`-backed token counter that the packaging already promises: a
counter implementing the existing `firewall/token_counting.py` seam, installed via
`pip install weaver-kernel[tiktoken]`, so context budgets count real model tokens
instead of character/byte heuristics.

## Why this matters

Budgets are the firewall's scalability promise to the LLM context window, and
context windows are measured in tokens. The extra is *already declared* in
`pyproject.toml` (`tiktoken = ["tiktoken>=0.6"]`) and the protocol seam exists with
a docstring naming tiktoken as the intended example — but installing the extra today
buys nothing, which is a small broken promise to anyone who reads the metadata.
Accurate counting makes budget downgrades fire at the right thresholds for real
deployments.

## Current evidence

- `pyproject.toml:68`: `tiktoken = ["tiktoken>=0.6"]` — declared, unused anywhere in `src/`.
- `firewall/token_counting.py:4`: docstring says the module exists to plug "token counters (for example, a `tiktoken`-based one)" into the firewall.
- `firewall/budgets.py`/`transform.py` currently drive decisions off byte-size estimates (see ISSUE 37).

## External context

tiktoken is the de-facto tokenizer for OpenAI-family models; Anthropic counting
differs — the design should name the encoding explicitly rather than guess per
model.

## Proposed implementation

1. Add `firewall/token_counting_tiktoken.py` (lazy import; helpful `ImportError`
   message naming the extra) implementing the existing counter protocol with a
   configurable encoding (default `cl100k_base` or `o200k_base` — decide and
   document).
2. Wire selection: `Firewall(token_counter=...)` already-shaped seam; document
   construction.
3. Cache encoder instances (expensive to build); counting must stay deterministic.
4. Bare-install safety: the module must not import tiktoken at package import time
   (pairs with ISSUE 19's no-extras CI job).

## AI-agent execution notes

- Inspect first: `firewall/token_counting.py` (protocol), `firewall/transform.py`/`budgets.py` (consumption), `otel.py` (pattern for optional-extra lazy import), `tests/test_firewall.py`.
- Edge cases: non-string data (count serialized form? document), very large strings (chunked counting), unknown encoding names (typed error).
- Follow the existing optional-dependency seam pattern exactly (mcp/otel precedents).

## Acceptance criteria

- With the extra installed, budgets use real token counts (tested with known strings and expected counts).
- Without the extra, behavior is unchanged and importing the public API does not require tiktoken.
- Helpful error when explicitly requesting the tiktoken counter without the extra.

## Test plan

Marked tests that skip without the extra; CI job (dev extras include it or add it);
bare-install job asserts no import-time dependency. Run `make ci`.

## Documentation plan

`docs/context_firewall.md` budgets section; README extras table; CHANGELOG `Added`.

## Migration and compatibility notes

Opt-in; defaults unchanged. Budget thresholds calibrated for byte counts may need
retuning when switching to token counts — document the difference.

## Risks and tradeoffs

tiktoken is a heavyweight binary dependency (hence extra-only); encoding choice can
mislead for non-OpenAI models — explicit configuration and docs mitigate.

## Suggested labels

ai, llm, product, performance

### PR DESCRIPTION
**Pull request by dgenio**
Closes #207, closes #211, closes #218, closes #174.

This is the grouped firewall PR identified during triage: four issues that share one implementation path — how the context firewall **measures, bounds, and faithfully represents** tool-output payloads before they reach the LLM. They live in `firewall/` + `handles.py` and are cleaner together (a single shared size estimator) than apart.

## What changed

- **`firewall/size_estimate.py` (new) — #207.** Allocation-free `estimated_size(value, *, limit=None)` that walks a structure to approximate its serialised length instead of `json.dumps`-ing it. `transform.py` now uses it for the raw-mode budget check (and drops `import json`), so a multi-MB raw payload is no longer fully serialised purely to measure it. Deterministic; exact for scalar containers, bounded error otherwise (only threshold comparisons depend on it).
- **`handles.py` — #211.** `HandleStore` gains opt-in `max_total_bytes` (evict oldest-first, never the just-stored entry) and `max_entry_bytes` (reject an over-cap payload with the new `HandleTooLarge` error rather than truncating it — keeps expansion faithful). Both default to `None` → behaviour unchanged. `current_bytes` exposes resident usage; eviction/expiry paths maintain the accounting.
- **`handle_query.py` (new) — refactor.** The `HandleStore.expand` query pipeline (validation → filter → paginate → project → redact → Frame) moves **verbatim** into a free `expand_handle()`; the store keeps a thin delegator that injects its TTL-checked fetch via a `lambda`. This was required to stay within the shrink-only module-size ratchet after adding the byte-budget code, and it brings `handles.py` from 371 → 264 lines (also chips at #188). `handles.py` is removed from `_SIZE_RATCHET` per the documented "shrank below 300 → drop it" rule.
- **`firewall/token_counting_tiktoken.py` (new) — #218.** `make_tiktoken_counter(encoding="cl100k_base")` implements the existing `TokenCounter` seam with a real tokenizer: lazy `import tiktoken` (base install stays light), cached encoder, eager resolution so a missing extra (`ImportError`) / unknown encoding (`FirewallError`) fails at construction. Exported from `weaver_kernel.firewall`; wired via `BudgetManager(token_counter=...)`.
- **`firewall/summarize.py` — #174.** Boolean columns are reported as `True`/`False` counts instead of being averaged (a `bool` is an `int` subclass — "mean of `is_active` = 0.7" is nonsense); truncated fact lists end in an explicit `… (N more facts omitted; full data via handle)` marker, never silently.
- **Docs/metadata:** `docs/context_firewall.md` (budget estimation, byte budgets, tiktoken usage, summary semantics), `CHANGELOG.md`, `HandleTooLarge` exported, `tiktoken.*` mypy override.

## Why

Grouped per the triage: same module, same bug category (size/budget accounting), one shared estimator. Decisions and assumptions (Mode B):
- **tiktoken default encoding `cl100k_base`** (broadest compatibility; `o200k_base` documented for GPT-4o/o-series) — the encoding is explicit, not guessed per model.
- **Oversized handle entry → reject** (not silent truncate), because a truncated raw dataset would corrupt expansion and the goal is bounding resident raw data.
- The issue's reference to `Firewall(token_counter=…)` is imprecise — the real seam is `BudgetManager(token_counter=…)`; wired there.
- tiktoken is **not** added to `[dev]`: it downloads BPE vocabularies over the network on first use, which must not be a test dependency. The counter's logic is unit-tested against a fake encoder injected into `sys.modules`, fully offline.

## How verified

- `make ci` → **EXIT 0**: `ruff format --check` (117 files), `ruff check` (all passed), `mypy --strict`, `pytest` **762 passed, 1 skipped** (branch-coverage `fail_under` gate met), all 13 examples run.
- New tests: `tests/test_size_estimate.py` (scalars, exactness on scalar containers, ±25% bounded error corpus, early-exit, non-serialisable fallback, deep-nesting no-recursion), `tests/test_token_counting_tiktoken.py` (fake-encoder: value handling, unknown-encoding `FirewallError`, missing-extra `ImportError`, `BudgetManager` wiring). Extended `tests/test_firewall.py` (bool-not-averaged, numeric still averaged, truncation marker present/absent, exactly-`max_facts` no false marker) and `tests/test_handles.py` (per-entry reject, total-bytes eviction order, evicted-handle expand → `HandleNotFound`, accounting release, invalid-config). Added `size_estimate` to the doctest gate.
- Behaviour-preserving: the existing `test_handles.py` expand suite (now exercising the delegator → `handle_query`) passes unchanged.

## Risks / caveats

- `estimated_size` is approximate; it preserves the only size-driven decision (raw-mode warning) and is bounded by tests, but it is not byte-identical to `json.dumps`.
- `max_entry_bytes` rejection raises from `store()`; only callers that opt in are affected (kernel default is `None`, so the invoke path is untouched). Kernel-level wiring of byte budgets is intentionally out of scope (HandleStore-only, per #211).
- One PR spans four issues by request; the `handle_query` move is verbatim but does enlarge the diff.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01HjetuNJX1EUJ2eVGfoLyHt

---
_Generated by [Claude Code](https://claude.ai/code/session_01HjetuNJX1EUJ2eVGfoLyHt)_

### CODE REVIEW

**Comment by dgenio** [comment]
Thanks — all five points were valid and are fixed in `25b26dd`:

- **`summarize.py` (dict & plain-list O(n) on the hot path):** both now iterate at most `max_facts` items via `islice` and derive the omitted count from the known length, so a huge payload no longer materialises one fact per item.
- **`handles.py` (store could stay over `max_total_bytes`):** a single entry larger than the binding cap (`max_entry_bytes` *or* `max_total_bytes`) is now rejected with `HandleTooLarge`, so `current_bytes` can never exceed the budget. The size check uses `estimated_size(..., limit=...)` so an over-cap payload isn't fully walked. Added `test_entry_larger_than_total_budget_is_rejected` / `test_total_budget_never_exceeded_after_stores`.
- **`size_estimate.py` (cyclic input):** the walk now records each container's id and visits it once, so self-referential structures terminate instead of hanging (covered by new cyclic-input tests).
- **`size_estimate.py` (traversal-order docstring):** the total is order-independent, so I corrected the docstring rather than reverse the push order (the early-exit value is documented as "only guaranteed `> limit`").

`make ci` is green locally (766 passed).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/weaver_kernel/firewall/summarize.py [hunk 1]
```diff
@@ -62,30 +92,32 @@ def _summarize_list_of_dicts(rows: list[dict[str, Any]], *, max_facts: int) -> l
...
 
-    return facts[:max_facts]
+    return _truncate_facts(facts, max_facts)
 
 
 def _summarize_dict(data: dict[str, Any], *, max_facts: int) -> list[str]:
     facts: list[str] = [f"Keys: {', '.join(sorted(data.keys())[:20])}"]
-    for k, v in list(data.items())[: max_facts - 1]:
+    for k, v in data.items():
         if isinstance(v, (int, float)):
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/firewall/summarize.py:132]: Fixed in `25b26dd`: `_summarize_dict()` now iterates at most `max_facts` items via `islice(data.items(), max_facts)` (`summarize.py:131`) and derives the omitted count from `total=1 + len(data)` passed to `_truncate_facts`. A large dict no longer materialises one fact per key on the firewall hot path, so `max_facts` is restored as a real bound.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/weaver_kernel/firewall/summarize.py [hunk 2]
```diff
@@ -96,16 +128,13 @@ def _summarize_dict(data: dict[str, Any], *, max_facts: int) -> list[str]:
...
+    return _truncate_facts(facts, max_facts)
 
 
 def _summarize_plain_list(data: list[Any], *, max_facts: int) -> list[str]:
     facts = [f"List of {len(data)} items"]
-    for item in data[: max_facts - 1]:
-        facts.append(repr(item)[:100])
-    return facts[:max_facts]
+    facts.extend(repr(item)[:100] for item in data)
+    return _truncate_facts(facts, max_facts)
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/firewall/summarize.py]: Fixed in `25b26dd`: `_summarize_plain_list()` now `repr()`s only `islice(data, max_facts)` elements (`summarize.py:149`). The true length is surfaced by the `List of N items` header and threaded to the omission marker via `total=1 + len(data)`, so there is no longer a full O(n) `repr()` walk before truncation.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/weaver_kernel/handles.py [hunk 3]
```diff
@@ -61,7 +86,18 @@ def store(
...
+            HandleTooLarge: If ``max_entry_bytes`` is set and the estimated size
+                of *data* exceeds it.
         """
+        size = estimated_size(data)
+        if self._max_entry_bytes is not None and size > self._max_entry_bytes:
+            raise HandleTooLarge(
+                f"Handle data for '{capability_id}' is ~{size} bytes, exceeding the "
+                f"per-entry cap of {self._max_entry_bytes} bytes; not stored."
+            )
+
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/handles.py]: Fixed in `25b26dd`: a single entry larger than the binding cap — `min(max_entry_bytes, max_total_bytes)` — is now rejected with `HandleTooLarge` (`handles.py:102-113`), so `current_bytes` can never end up above `max_total_bytes`. The check uses `estimated_size(data, limit=reject_at)` so an over-cap payload that will be rejected anyway isn't fully walked. Covered by `test_entry_larger_than_total_budget_is_rejected` and `test_total_budget_never_exceeded_after_stores`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/weaver_kernel/firewall/size_estimate.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        elif isinstance(cur, dict):
+            n = len(cur)
+            total += _CONTAINER_DELIMS + max(0, n - 1) * _ITEM_SEP
+            for key, val in cur.items():
+                total += len(str(key)) + _QUOTES + _KV_SEP
+                stack.append(val)
+        elif isinstance(cur, (list, tuple)):
+            n = len(cur)
+            total += _CONTAINER_DELIMS + max(0, n - 1) * _ITEM_SEP
+            stack.extend(cur)
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/firewall/size_estimate.py:106]: Addressed in `25b26dd` by correcting the docstring rather than reversing the push order. `estimated_size` no longer claims native/insertion order: the summed total is documented as order-independent (`size_estimate.py:11-12`), and the `limit` early-exit is documented as only guaranteeing a return value `> limit`, with which member trips the threshold left unspecified (`size_estimate.py:54-56`). Because the total is genuinely order-independent, the only defect was the stale docstring claim, which is now gone.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/weaver_kernel/firewall/size_estimate.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        >>> estimated_size(None)
+        4
+        >>> estimated_size("hi")
+        4
+        >>> estimated_size([1, 2, 3])
+        9
+    """
+    total = 0
+    stack: list[Any] = [value]
+    while stack:
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/weaver_kernel/firewall/size_estimate.py:77]: Fixed in `25b26dd`: the walk now records each container's `id()` in a `seen` set and skips already-counted containers (`size_estimate.py:71`, `:88-90`, `:97-99`), so self-referential / cyclic inputs terminate instead of hanging. This closes the DoS gap relative to the old `json.dumps` path; new cyclic-input tests in `tests/test_size_estimate.py` cover it.

---
_Generated by [Claude Code](https://claude.ai/code)_
